src/reader/mod.rs

```diff
@@ -1,18 +1,139 @@
 //! A module to handle `Reader`
 
 use std::borrow::Cow;
-use std::io::{self, BufRead, BufReader};
-use std::{fs::File, path::Path, str::from_utf8};
+use std::str::from_utf8;
 
 #[cfg(feature = "encoding")]
 use encoding_rs::{Encoding, UTF_16BE, UTF_16LE, UTF_8};
 
 use crate::errors::{Error, Result};
 use crate::events::{BytesCData, BytesDecl, BytesEnd, BytesStart, BytesText, Event};
-use crate::name::{LocalName, NamespaceResolver, QName, ResolveResult};
 
 use memchr;
 
+macro_rules! builder_methods {
```

Comment:
@dralley , you've asked, why I use macro here. Although it looks like you have deleted it, I will still answer.

The change in which this macro is introduced, removes `impl DerefMut<Target=Reader> for NsReader`. That means, that all `Reader`'s methods that accept `&mut self` won't be available for the `NsReader`, including the ones that implemented by this macro.

Because we need a way to configure our reader, we need to reimplement necessary methods that accept `&mut self`. In order to not duplicate the code and the doc of this configure methods, I've put them into the macro and embed to both readers.

Reply:
Yes, I figured that part out after a few minutes.

Reply:
Added that explanation to the commit message